### PR TITLE
Fix server removal

### PR DESCRIPTION
**Pull request by fortuna**
This should fix the server removal issue.

The problem is that `dom-repeat` always removes the last node from the DOM tree, regardless of what entry was removed from the serverList. This has always happened, but the old code would always set all the view fields upon selection, mitigating the problem. I didn't capture the bug when testing because removing the last added server always works fine. You will only see the problem when you add two servers and delete the one before the last.

The old behavior had issues though, like flashing old content before showing new, and corner cases.

The new code uses the [lit-html repeat directive](https://lit-html.polymer-project.org/guide/writing-templates#repeating-templates), which properly maps the list items to dom nodes, deleting the right node when the item is deleted. This required introducing a new element: `outline-server-list`.

Note that `getServerView` is now async. That's because we must wait for the component to be updated before a new view is available. Without the updated wait, we fail to get the view for a newly created server.

The deeper problem is that we are doing web components the wrong way. We should not be manipulating the views. The way we are doing, Polymer becomes unaware of the state changes. We should express all the state with properties instead, and the views should just be dumb representations of that state. In this case, all the ServerView parameters should be part of the ServerListEntry. This way the resulting DOM would have the right state, regardless of what node was removed.

Let's prioritize this change so we can release a fix before the weekend.

/cc @alalamav

### CODE REVIEW

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 1]
```diff
@@ -493,7 +493,7 @@ export class ServerView extends DirMixin(PolymerElement) {
                 <img class="digital-ocean-icon" src="images/do_white_logo.svg">
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:493]: This is unrelated cleanup

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 2]
```diff
@@ -623,38 +623,33 @@ export class ServerView extends DirMixin(PolymerElement) {
...
         myConnection: Object,
         totalInboundBytes: Number,
         baselineDataTransfer: Number,
-        accessKeyRows: {type: Array},
+        accessKeyRows: Array,
         hasNonAdminAccessKeys: Boolean,
         metricsEnabled: Boolean,
-        monthlyOutboundTransferBytes: {type: Number},
-        monthlyCost: {type: Number},
-        managedServerUtilzationPercentage: {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:647]: @JonathanDCohen FYI, we don't need to define computed properties, unless an external party needs the value.

File: src/server_manager/web_app/ui_components/outline-server-list.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    `)}</div>`;
+  }
+
+  async getServerView(serverId: string): Promise<ServerView> {
+    if (!serverId) {
+      return null;
+    }
+    // We need to wait updates to be completed or the view may not yet be there.
+    await this.updateComplete;
+    return this.shadowRoot.querySelector<ServerView>(`#${this.makeViewId(serverId)}`);
```
Comment (JonathanDCohen) [inline src/server_manager/web_app/ui_components/outline-server-list.ts:53]: nit: split the id out into a variable to reduce the line noise

File: src/server_manager/web_app/ui_components/outline-server-list.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  @property({type: Array}) serverList: ServerListEntry[];
+  @property({type: String}) selectedServerId: string;
+  @property({type: Function}) localize: Function;
+  @property({type: String}) language: string;
+
+  render() {
+    if (!this.serverList) {
+      return;
+    }
+    return html`<div>${repeat(this.serverList, e => e.id, e => html`
```
Comment (JonathanDCohen) [inline src/server_manager/web_app/ui_components/outline-server-list.ts:37]: Let's add typing for the list elements?

File: src/server_manager/web_app/app.ts [hunk 5]
```diff
@@ -876,9 +876,10 @@ export class App {
 
   private addAccessKey() {
```
Comment (JonathanDCohen) [inline src/server_manager/web_app/app.ts:877]: Let's use async-await for this whole function instead of mixing styles

File: src/server_manager/web_app/ui_components/app-root.js [hunk 6]
```diff
@@ -658,16 +656,10 @@ export class AppRoot extends mixinBehaviors
...
-  getServerView(displayServerId) {
-    if (!displayServerId) {
-      return null;
-    }
-    // Render to ensure that the server view has been added to the DOM.
-    this.$.serverView.querySelector('dom-repeat').render();
-    const selectedServerId = this._base64Encode(displayServerId);
-    return this.$.serverView.querySelector(`#serverView-${selectedServerId}`);
+  async getServerView(displayServerId) {
+    return await this.shadowRoot.querySelector('#serverView').getServerView(displayServerId);
```
Comment (JonathanDCohen) [inline src/server_manager/web_app/ui_components/app-root.js:662]: Do we need to await here?  We can just return the promise since callers have to await to use it

File: src/server_manager/web_app/ui_components/outline-server-list.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 The Outline Authors
```
Comment (JonathanDCohen) [inline src/server_manager/web_app/ui_components/outline-server-list.ts:1]: @mpmcroy this is another case where I think Vini would run into the same problem I did with the dialog if he wanted to write unit tests for this element.  Let's put our heads together and figure this out so we can increase our test coverage as we migrate the UI to TS / LitElement?

File: src/server_manager/web_app/ui_components/app-root.js [hunk 8]
```diff
@@ -658,16 +656,10 @@ export class AppRoot extends mixinBehaviors
...
    * @param {string} displayServerId
-   * @returns {ServerView}
+   * @returns {Promise<ServerView>}
    */
-  getServerView(displayServerId) {
-    if (!displayServerId) {
-      return null;
-    }
-    // Render to ensure that the server view has been added to the DOM.
-    this.$.serverView.querySelector('dom-repeat').render();
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/app-root.js:668]: Wow, thanks for finding the root cause. This seems like a very tricky bug.

There is still one thing that I don't understand. Why doesn't forcing a re-render address (side-step) the issue? I would expect it to rebind all of the views regardless of how the serverList items were added/removed.

File: src/server_manager/web_app/ui_components/outline-server-list.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  @property({type: Function}) localize: Function;
+  @property({type: String}) language: string;
+
+  render() {
+    if (!this.serverList) {
+      return;
+    }
+    return html`<div>${repeat(this.serverList, e => e.id, e => html`
+      <outline-server-view
+        .id="${this.makeViewId(e.id)}"
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/outline-server-list.ts:39]: I kind of like how this is an implementation detail now.

**Review by mpmcroy — APPROVED** [review]

File: src/server_manager/web_app/app.ts (hunk 5 above)
Comment (fortuna) [reply src/server_manager/web_app/app.ts]: Done

File: src/server_manager/web_app/ui_components/app-root.js (hunk 6 above)
Comment (fortuna) [reply src/server_manager/web_app/ui_components/app-root.js:662]: await makes it clearer.   Otherewise one may think it may return a promise with a promise, even though that doesn't really happen.

File: src/server_manager/web_app/ui_components/app-root.js (hunk 8 above)
Comment (fortuna) [reply src/server_manager/web_app/ui_components/app-root.js:668]: This is what happens:
serverList is:
- A
- B
- C

When you delete B, the dom for the server views will look like:
- id: A, properties A
- id: C, properties B

That's because the id, used to lookup the view, is bound to the serverList. But all the other properties are set directly on the components, bypassing the model binding. The rebinding works, for the properties that are bound. That's why I said we are doing it wrong. The entire state should be bound. That could be done with a DisplayServer kind of object that we bind.

We don't have the problem with the access keys because we use the AccessKeyRow for binding.

File: src/server_manager/web_app/ui_components/outline-server-list.ts (hunk 4 above)
Comment (fortuna) [reply src/server_manager/web_app/ui_components/outline-server-list.ts:37]: serverList is already typed, so are the parameters of repeat.

File: src/server_manager/web_app/ui_components/outline-server-list.ts (hunk 3 above)
Comment (fortuna) [reply src/server_manager/web_app/ui_components/outline-server-list.ts]: Done

**Review by fortuna — COMMENTED** [review]
Another look?

**Review by JonathanDCohen — APPROVED** [review]

File: src/server_manager/web_app/ui_components/app-root.js (hunk 8 above)
Comment (mpmcroy) [reply src/server_manager/web_app/ui_components/app-root.js:668]: Thanks, that clears things up.

What I take away from this is that we're mixing MVVM and MVP. Some properties are bound through the ViewModel (`ServerListEntry`) and others are bound directly via the presenter (`app.ts`). I think this would be great to flesh out when we refactor parts of the codebase into LitElement apps.

Comment (fortuna) [reply src/server_manager/web_app/ui_components/app-root.js:668]: For the most part we should be binding via the webcomponent model. Dealing with direct property setting and the views can tie us to a specific layout and is a pain.
